python/ray/air/result.py

```diff
@@ -110,3 +117,109 @@ def _repr(self, indent: int = 0) -> str:
 
     def __repr__(self) -> str:
         return self._repr(indent=0)
+
+    @staticmethod
+    def _validate_trial_dir(trial_dir: str):
+        """Check the validity of the local trial folder."""
+
+        assert os.path.exists(trial_dir), f"Trail folder {trial_dir} doesn't exists!"
+
+        all_files = os.listdir(trial_dir)
+        for file in Result._restore_required_files:
+            assert file in all_files, f"{file} not found in the trial folder!"
+
+    @classmethod
+    def from_path(cls, path: str) -> "Result":
+        """Restore a Result object from local trial directory.
+
+        Args:
+            path: the path to a local trial directory.
+
+        Returns:
+            A :py:class:`Result` object of that trial.
+        """
+
+        # TODO(yunxuanx): restoration from cloud storage
+        local_path = path
+
+        cls._validate_trial_dir(local_path)
+
+        # Restore reported metrics
+        with open(Path(local_path) / EXPR_RESULT_FILE, "r") as f:
+            json_list = [json.loads(line) for line in f if line]
+            metrics_df = pd.json_normalize(json_list, sep="/")
+
+        metrics = metrics_df.iloc[-1].to_dict() if not metrics_df.empty else {}
+
+        # Restore all checkpoints from the checkpoint folders
+        ckpt_dirs = [
+            os.path.join(local_path, entry)
+            for entry in os.listdir(local_path)
+            if entry.startswith("checkpoint_")
+        ]
+
+        if ckpt_dirs:
+            checkpoints = [
+                Checkpoint.from_directory(ckpt_dir) for ckpt_dir in ckpt_dirs
+            ]
+
+            checkpoint_metrics = [
+                metrics_df[metrics_df[TRAINING_ITERATION] == ckpt.iteration].to_dict(
+                    "records"
+                )[0]
+                for ckpt in checkpoints
+            ]
+
+            # TODO(air-team): make metrics a property of Checkpoint
+            best_checkpoints = list(zip(checkpoints, checkpoint_metrics))
+            latest_checkpoint = max(checkpoints, key=lambda ckpt: ckpt.id)
+        else:
+            best_checkpoints = latest_checkpoint = None
+
+        # Restore the trial error if it exists
+        error = None
+        error_file_path = Path(local_path) / EXPR_ERROR_PICKLE_FILE
+        if os.path.exists(error_file_path):
+            error = pickle.load(open(error_file_path, "rb"))
+
+        return Result(
+            metrics=metrics,
+            checkpoint=latest_checkpoint,
+            _local_path=local_path,
+            _remote_path=None,
+            metrics_dataframe=metrics_df,
+            best_checkpoints=best_checkpoints,
+            error=error,
+        )
+
+    @PublicAPI(stability="alpha")
+    def get_best_checkpoint(self, metric: str, mode: str) -> Optional[Checkpoint]:
```

Comment:
With ResultGrid, we're able to keep track of the metric and mode passed in through the `TuneConfig`. Now, with only the information in a single trial directory, we don't currently keep that metric/mode around, so a user must pass it in. 

Should we save that info around?

Reply:
I think this should be considered together with making metrics a property of Checkpoint. 

For example, saving a json file in the checkpoint folder with something like:

```
{"checkpoint_score_attribute": "abc", "checkpoint_score_order": "max", "latest_metrics": {"abc": 123, "xyz": 456}}
```

I think we can open a separate PR for this, since it introduces more code changes.